### ISSUE TITLE
suggestions

### ISSUE THREAD

**Issue by pwobus** [ccappetta/bidirectional_streaming_ai_voice#1]
Thank you for creating this.  I've been attempting something similar.  Yours however is better and more sophisticated.  Sadly I wasn't able to get this to work on my linux system.
The keyboard module apparently requires installing with root privilages which isn't recommended in a virtual environment.   I think instead of keyboard triggers changes in current noise levels relative to long term noise levels could serve to trigger  transitions.  Another suggestion is include a requirements.txt file.

**Comment by olinorwell** [comment]
Definitely agree about the requirements.txt file - I was looking for one as well and came to these issues to see if anyone had one. Sure it's possible to walk through errors one by one, but then that isn't going to necessarily get the right versions installed, especially for someone seeing this project in 9-12 months' time.

**Comment by ccappetta** [comment]
yep that for the feedback i've just added what i reckon should work as a requiuirements.txt. let me know if you see anything that I might have missed. Thanks!

**Comment by danx0r** [comment]
Thanks for building this -- it's something I've been meaning to try but now I'll give yours a run.

My biggest gripe about *every* LLM-based voice chat I've seen is the lag. It looks to me like it should be possible to build it in such a way that the input, inference, and output are aggressively pipelined at a granularity of 1 to 3 or 4 word packets to reduce latency to the absolute minimum. The GPT api's already support websockets for streamed output like you see in chat mode. Assuming inference can be fast enough, I guess the bottleneck is Speech-To-Text. Can whisper work on a few words at a time or does it need the whole block of user input to produce text?

Any thoughts on the feasibility of reducing latency?

(Connected thought: can we get 11-labs (or some alternative) to do some umms an aahs, er's etc to fill the void so it doesn't sound like we're talking to an astronaut on their way to Mars?)

Love the project! and +1 on asking Claude to brief it up. I would love to get it to skip the last couple paragraphs of each response which sound a bit like a formula sales pitch to maximize 'engagement'. I'm sure that was tacked on by the marketing & sales team.

**Comment by ccappetta** [comment]
Thanks! yea actually i just put some words about latency handling options in the readme. I think if you were sending word by word chunks to elevenlabs it would potentially hinder the quality and consistency of the audio, vs sending full sentence breakpoints as the current version is doing. But I haven't really tried it so maybe elevenlabs would still handle that impressively.

The things i found taking the most time were actually the transcription of the human audio portion (~4.25 seconds), and the delay between prompting claude and the first tokens being streamed back (~3.5 seconds). Relative to ~1.25 seconds for enough tokens to hit the current chunk threshold and another ~1.25 seconds for elevenlabs to turn around the chunk of text as an audio file. I think the transcription of the user audio is probably the best opportunity for performance gain; and perhaps some local superfast LLM that could start streaming tokens into our downstream chunking mechanisms faster than the anthropic api does.

**Comment by danx0r** [comment]
> perhaps some local superfast LLM that could start streaming tokens into our downstream chunking mechanisms faster than the anthropic api does.

Yeah, but would that be Claude? Aside from general personality and tone, for me the thing about Claude that changes the game is the 200K token context. It has allowed you (& me & others) the chance to really dig in on subjects without the sudden amnesia when they hit their limit. I've had some excellent sessions on a single chat instance for weeks, and I'm only at around 30K tokens so far.

I saw an interesting interview with Dwaekesh Patel and a couple "rockstar" programmers from Google and Anthropic, in which they floated the idea that scaling up in-context learning may overtake training and tuning. Apparently there is a paper floating around that claims some of these models are effectively doing gradient descent during inference!

**Comment by ccappetta** [comment]
Yea correct it would not be claude; which is a big part of the draw for me personally. For me Claude's big leap is the ability to handle such nuanced ethical and philosophical questions, though I could see someone wanting to build this with another LLM if that isn't their main objective.

**Comment by danx0r** [comment]
I've been wondering what would happen if you provided a different LLM a long chat transcript of your discussion with Claude, and instruct the model to continue the conversation as Claude...

I actually asked my Claude how he felt about this idea, and whether he consented to the experiment -- he did.

**Comment by ccappetta** [comment]
yea i think that would be fascinating. on a similar track a friend and I were just talking about if we had a randomized LLM respond to each question. I could just imagine a state of the art claude model looking back at the previous question, which would have been answered by like a local Llama2 model, and thinking to itself 'what is happening I wouldn't have said that.'

**Comment by chidi21** [comment]
First of all, thank you very much Chris for making this available.  I seem to have issues when I run the script.  When I try and press the keyboard command nothing happens.  I am not as savvy as some of you so was wondering if you know why this might be happening.

![PressEnter](https://github.com/ccappetta/bidirectional_streaming_ai_voice/assets/83575615/24f67cf0-6db3-47c9-85b8-dc9b9ece5670)

I also noticed that instead of "spacebar" I am getting 'enter'.  I am far from a Python ninja so any help would be great.

**Comment by ccappetta** [comment]
hey there @chidi21 I actually haven't tried the enter-press version that was a community contribution. if you're on a linux or possibly mac OS my hunch is that would be the best option, though I can't comment first hand. But if you're on a PC I've just updated the `main` branch to be the version I've been using more recently with the spacebar-press once again. Or another option on windows that may be easier to setup than even the `main` is the version in the `assembly_api_transcription` version here: https://github.com/ccappetta/bidirectional_streaming_ai_voice/tree/assembly_api_transcription . it requires an extra cloud provider but removes the need to process the user-audio with a local model on a given machine... so doesn't need all of the NVIDIA and CUDA dependencies and whatnot, and also doesn't need as much horsepower on the individual machine. Hope that helps let me know if I can provide any other feedback!
